mypyc/genops.py

```diff
@@ -382,26 +382,56 @@ def __init__(self, fitem: FuncItem = INVALID_FUNC_DEF, name: str = '',
         self.fitem = fitem
         self.name = name
         self.ns = namespace
-        # Callable classes are ClassIR instances implementing the '__call__' method, used to
-        # represent functions that are nested inside of other functions.
-        self.callable_class = INVALID_CLASS
+        # Callable classes implement the '__call__' method, and are used to represent functions
+        # that are nested inside of other functions.
+        self.callable_class = CallableClass()
         # Environment classes are ClassIR instances that contain attributes representing the
         # variables in the environment of the function they correspond to. Environment classes are
         # generated for functions that contain nested functions.
         self.env_class = INVALID_CLASS
-        # The register associated with the 'self' instance for function classes.
-        self.self_reg = INVALID_VALUE  # type: Value
-        # Environment class registers are the local registers associated with instances of an
-        # environment class, used for getting and setting attributes. env_reg is the register
-        # associated with the current environment, and prev_env_reg is the self.__mypyc_env__ field
-        # associated with the previous environment.
-        self.env_reg = INVALID_VALUE  # type: Value
-        self.prev_env_reg = INVALID_VALUE  # type: Value
         # These are flags denoting whether a given function is nested or contains a nested
         # function.
+        # Generator classes implement the '__next__' method, and are used to represent generators
+        # returned by generator functions.
+        self.generator_class = GeneratorClass()
```

Comment:
I'm not sure that I love always having a `CallableClass` and a `GeneratorClass` allocated for each function, since usually they don't exist, right? It feels a bit clunky.

Reply:
It might be best to make them `Optional` and that replace `is_nested` and `is_generator`. (Or make those just be properties that check if the `generator_class` and `callable_class` are None)